### PR TITLE
Apply azure container instance SDK 2017-10-01

### PR DESCRIPTION
**Pull request by yolocs**
## Description

---

This checklist is used to make sure that common guidelines for a pull request are followed. You can find a more complete discussion of PowerShell cmdlet best practices [here](https://msdn.microsoft.com/en-us/library/dd878270(v=vs.85).aspx).

- [x] **I have read the [contribution guidelines](https://github.com/Azure/azure-powershell/blob/preview/CONTRIBUTING.md).**
- [x] **If changes were made to any cmdlet, the XML help was regenerated using the [platyPSHelp module](https://github.com/Azure/azure-powershell/blob/preview/documentation/help-generation.md).**
- [x] **If any large changes are made to a service, they are reflected in the respective [change log](https://github.com/Azure/azure-powershell/blob/preview/CONTRIBUTING.md#updating-the-change-log).**

### [General Guidelines](https://github.com/Azure/azure-powershell/blob/preview/CONTRIBUTING.md#general-guidelines)
- [x] Title of the pull request is clear and informative.
- [x] There are a small number of commits, each of which have an informative message. This means that previously merged commits do not appear in the history of the PR. For more information on cleaning up the commits in your PR, [see this page](https://github.com/Azure/azure-powershell/blob/preview/documentation/cleaning-up-commits.md).
- [x] The pull request does not introduce [breaking changes](https://github.com/Azure/azure-powershell/blob/preview/documentation/breaking-changes/breaking-changes-definition.md) (unless a major version change occurs in the assembly and module).

### [Testing Guidelines](https://github.com/Azure/azure-powershell/blob/preview/CONTRIBUTING.md#testing-guidelines)
- [x] Pull request includes test coverage for the included changes.
- [x] PowerShell scripts used in tests should do any necessary setup as part of the test or suite setup, and should not use hard-coded values for locations or existing resources.

### [Cmdlet Signature Guidelines](https://github.com/Azure/azure-powershell/blob/preview/CONTRIBUTING.md#cmdlet-signature-guidelines)
- [x] New cmdlets that make changes or have side effects should implement `ShouldProcess` and have `SupportShouldProcess=true` specified in the cmdlet attribute. You can find more information on `ShouldProcess` [here](https://github.com/Azure/azure-powershell/wiki/PowerShell-Cmdlet-Design-Guidelines#supportsshouldprocess).
- [x] Cmdlet specifies `OutputType` attribute if any output is produced - if the cmdlet produces no output, it should implement a `PassThru` parameter.

### [Cmdlet Parameter Guidelines](https://github.com/Azure/azure-powershell/blob/preview/CONTRIBUTING.md#cmdlet-parameter-guidelines)
- [x] Parameter types should not expose types from the management library - complex parameter types should be defined in the module.
- [x] Complex parameter types are discouraged - a parameter type should be simple types as often as possible. If complex types are used, they should be shallow and easily creatable from a constructor or another cmdlet.
- [x] Cmdlet parameter sets should be mutually exclusive - each parameter set must have at least one mandatory parameter not in other parameter sets.

### CODE REVIEW

**Comment by cormacpayne** [comment]
@azuresdkci test this please

**Comment by cormacpayne** [comment]
@yolo3301 would you mind resolving the merge conflicts?

File: src/ResourceManager/ContainerInstance/AzureRM.ContainerInstance.psd1 [hunk 1]
```diff
@@ -12,7 +12,7 @@
 # RootModule = ''
 
 # Version number of this module.
-ModuleVersion = '0.1.0'
+ModuleVersion = '0.1.1'
```
Comment (cormacpayne) [inline src/ResourceManager/ContainerInstance/AzureRM.ContainerInstance.psd1]: @yolo3301 you can revert any change made to versions in this PR. We will handle the versioning before publishing the module

File: src/ResourceManager/ContainerInstance/ChangeLog.md [hunk 2]
```diff
@@ -19,6 +19,12 @@
 -->
 ## Current Release
 
+## Version 0.1.1
```
Comment (cormacpayne) [inline src/ResourceManager/ContainerInstance/ChangeLog.md]: @yolo3301 please remove the `## Version 0.1.1` header and move the below content to be under `## Current Release`

File: src/ResourceManager/ContainerInstance/Commands.ContainerInstance.Test/Properties/AssemblyInfo.cs [hunk 3]
```diff
@@ -32,5 +32,5 @@
 // You can specify all the values or you can default the Build and Revision Numbers
 // by using the '*' as shown below:
 // [assembly: AssemblyVersion("1.0.*")]
-[assembly: AssemblyVersion("0.1.0")]
-[assembly: AssemblyFileVersion("0.1.0")]
+[assembly: AssemblyVersion("0.1.1")]
```
Comment (cormacpayne) [inline src/ResourceManager/ContainerInstance/Commands.ContainerInstance.Test/Properties/AssemblyInfo.cs]: @yolo3301 please revert the changes made in this file

File: src/ResourceManager/ContainerInstance/Commands.ContainerInstance/Commands/NewAzureContainerGroupCommand.cs [hunk 4]
```diff
@@ -116,9 +144,9 @@ public class NewAzureContainerGroupCommand : ContainerInstanceCmdletBase
 
         [Parameter(
             Mandatory = false,
-            HelpMessage = "The port to open. Default: 80")]
+            HelpMessage = "The ports to open. Default: [80]")]
         [ValidateNotNullOrEmpty]
-        public int? Port { get; set; }
+        public int[] Ports { get; set; }
```
Comment (cormacpayne) [inline src/ResourceManager/ContainerInstance/Commands.ContainerInstance/Commands/NewAzureContainerGroupCommand.cs]: @yolo3301 why was this change made?

File: src/ResourceManager/ContainerInstance/Commands.ContainerInstance/Properties/AssemblyInfo.cs [hunk 5]
```diff
@@ -32,5 +32,5 @@
 // You can specify all the values or you can default the Build and Revision Numbers
 // by using the '*' as shown below:
 // [assembly: AssemblyVersion("1.0.*")]
-[assembly: AssemblyVersion("0.1.0")]
```
Comment (cormacpayne) [inline src/ResourceManager/ContainerInstance/Commands.ContainerInstance/Properties/AssemblyInfo.cs]: @yolo3301 please revert the changes made in this file

File: src/ResourceManager/ContainerInstance/Commands.ContainerInstance/help/New-AzureRmContainerGroup.md [hunk 6]
```diff
@@ -122,12 +162,92 @@ IpAddress                : 13.88.10.240
...
+State                    : Running
+Events                   : {}
+```
+
+This commands creates a container group that mounts the provided Azure File share to `/mnt/azfile`.
+
 ## PARAMETERS
 
+### -AzureFileVolumeAccountKey
+The storage account key of the Azure File share to mount.```yaml
```
Comment (cormacpayne) [inline src/ResourceManager/ContainerInstance/Commands.ContainerInstance/help/New-AzureRmContainerGroup.md]: @yolo3301 please make sure that the trailing ` ```yaml ` string is moved to a line by itself. This applies to all markdown files that have been edited in this file. This is caused from a bug in the `platyPS` module.

File: tools/AzureRM/AzureRM.psd1 [hunk 7]
```diff
@@ -65,7 +65,7 @@ RequiredModules = @(@{ModuleName = 'AzureRM.Profile'; RequiredVersion = '4.0.0';
                @{ModuleName = 'AzureRM.CognitiveServices'; RequiredVersion = '0.9.0'; }, 
                @{ModuleName = 'AzureRM.Compute'; RequiredVersion = '4.0.1'; }, 
                @{ModuleName = 'AzureRM.Consumption'; RequiredVersion = '0.3.0'; }, 
-               @{ModuleName = 'AzureRM.ContainerInstance'; RequiredVersion = '0.1.0'; }, 
+               @{ModuleName = 'AzureRM.ContainerInstance'; RequiredVersion = '0.1.1'; },
```
Comment (cormacpayne) [inline tools/AzureRM/AzureRM.psd1]: @yolo3301 please revert the change made in this file

File: src/ResourceManager/ContainerInstance/Commands.ContainerInstance/Commands/NewAzureContainerGroupCommand.cs (hunk 4 above)
Comment (yolo3301) [reply src/ResourceManager/ContainerInstance/Commands.ContainerInstance/Commands/NewAzureContainerGroupCommand.cs]: We support multiple ports. In Azure CLI we have already made a similar change.

**Comment by yolo3301** [comment]
@cormacpayne Created a simple [review page](https://github.com/yolo3301/azure-powershell/wiki/Feature-review-for-%22Apply-azure-container-instance-SDK-2017-10-01%22)

**Comment by maddieclayton** [comment]
Page for cmdlet review can now be found here: https://github.com/Azure/azure-powershell-pr/wiki/Apply-azure-container-instance-SDK-2017-10-01

File: src/ResourceManager/ContainerInstance/Commands.ContainerInstance/Commands/NewAzureContainerGroupCommand.cs (hunk 4 above)
Comment (cormacpayne) [reply src/ResourceManager/ContainerInstance/Commands.ContainerInstance/Commands/NewAzureContainerGroupCommand.cs]: @yolo3301 from PowerShell's _Strongly Encouraged Development Guidelines_ (the snippet can be found [here](https://github.com/Azure/azure-powershell/wiki/PowerShell-Cmdlet-Design-Guidelines#singular-noun-1)), parameter names should be singular _unless_ there is validation for the parameter preventing the user from passing only one object. If a user can pass just one port, then this parameter should be `public int[] Port`

Comment (yolo3301) [reply src/ResourceManager/ContainerInstance/Commands.ContainerInstance/Commands/NewAzureContainerGroupCommand.cs]: Good to know. Updated.

**Comment by cormacpayne** [comment]
@azuresdkci test this please

File: src/ResourceManager/ContainerInstance/Commands.ContainerInstance.Test/ScenarioTests/ContainerInstanceTests.ps1 [hunk 8]
```diff
@@ -23,17 +23,22 @@ function Test-AzureRmContainerGroup
     $location = Get-ProviderLocation "Microsoft.ContainerInstance/ContainerGroups"
     $image = "nginx"
     $osType = "Linux"
```
Comment (maddieclayton) [inline src/ResourceManager/ContainerInstance/Commands.ContainerInstance.Test/ScenarioTests/ContainerInstanceTests.ps1]: Please add a test for the new parameter set

File: src/ResourceManager/ContainerInstance/Commands.ContainerInstance/Commands/NewAzureContainerGroupCommand.cs [hunk 9]
```diff
@@ -46,6 +47,12 @@ public class NewAzureContainerGroupCommand : ContainerInstanceCmdletBase
             ParameterSetName = CreateContainerGroupWithRegistryParamSet,
             ValueFromPipelineByPropertyName = true,
             HelpMessage = "The resource group name.")]
+        [Parameter(
```
Comment (maddieclayton) [inline src/ResourceManager/ContainerInstance/Commands.ContainerInstance/Commands/NewAzureContainerGroupCommand.cs]: If a parameter is in all sets, and Mandatory and in the same position, you don't need to have a separate attribute for each parameter set.  Please erase all parameter attributes and put in:
```
[Parameter(
Mandatory = true,
Position = 0,
ValueFromPipelineByPropertyName = true,
HelpMessage = "The resource group name.")]
```
Apply this all other parameters also.  Especially if you plan to add more parameter sets in the future, there is no need to create a new attribute each time.

File: src/ResourceManager/ContainerInstance/Commands.ContainerInstance/Commands/NewAzureContainerGroupCommand.cs [hunk 10]
```diff
@@ -149,6 +177,34 @@ public class NewAzureContainerGroupCommand : ContainerInstanceCmdletBase
         [ValidateNotNullOrEmpty]
         public PSCredential RegistryCredential { get; set; }
 
+        [Parameter(
```
Comment (maddieclayton) [inline src/ResourceManager/ContainerInstance/Commands.ContainerInstance/Commands/NewAzureContainerGroupCommand.cs]: Quick recommendation: move these parameters above all of the non-mandatory parameters so that the users will see them more quickly when scanning help.

File: src/ResourceManager/ContainerInstance/Commands.ContainerInstance/Commands/NewAzureContainerGroupCommand.cs [hunk 11]
```diff
@@ -149,6 +177,34 @@ public class NewAzureContainerGroupCommand : ContainerInstanceCmdletBase
...
+            HelpMessage = "The storage account name of the Azure File share to mount.")]
+        [ValidateNotNullOrEmpty]
+        public string AzureFileVolumeAccountName { get; set; }
+
+        [Parameter(
+            Mandatory = true,
+            ParameterSetName = CreateContainerGroupWithAzureFileVolumeParamSet,
+            HelpMessage = "The storage account key of the Azure File share to mount.")]
+        [ValidateNotNullOrEmpty]
+        public PSCredential AzureFileVolumeAccountKey { get; set; }
```
Comment (maddieclayton) [inline src/ResourceManager/ContainerInstance/Commands.ContainerInstance/Commands/NewAzureContainerGroupCommand.cs]: Is account key the correct name for this parameter? I wouldn't usually consider a PSCredential to be a "key".  Please correct me if I am wrong.

File: src/ResourceManager/ContainerInstance/Commands.ContainerInstance/Models/PSContainer.cs [hunk 12]
```diff
@@ -62,7 +62,12 @@ public class PSContainer
         /// <summary>
         /// Gets or sets the events.
         /// </summary>
-        public IList<PSContainerEvent> Events { get; set; }
```
Comment (maddieclayton) [inline src/ResourceManager/ContainerInstance/Commands.ContainerInstance/Models/PSContainer.cs]: Is the change from PSContainerEvent to PSEvent an unrelated change? Just want to make sure I understand the scope of the changes.

File: src/ResourceManager/ContainerInstance/Commands.ContainerInstance/Models/PSContainerGroup.cs [hunk 13]
```diff
@@ -104,6 +104,16 @@ public string ResourceGroupName
         /// </summary>
         public IList<Volume> Volumes { get; set; }
 
+        /// <summary>
+        /// Gets or sets the state of the container group.
+        /// </summary>
+        public string State { get; set; }
```
Comment (maddieclayton) [inline src/ResourceManager/ContainerInstance/Commands.ContainerInstance/Models/PSContainerGroup.cs]: Same here.  Is this just to keep up to date with the current SDK?

File: src/ResourceManager/ContainerInstance/Commands.ContainerInstance/help/New-AzureRmContainerGroup.md [hunk 14]
```diff
@@ -122,12 +162,100 @@ IpAddress                : 13.88.10.240
 Ports                    : {80}
 OsType                   : Linux
 Volumes                  :
+State                    : Running
+Events                   : {}
 ```
 
 This commands creates a container group using a custom image from a custom container image registry.
 
+### Example 6: Creates a container group that mounts Azure File volume
```
Comment (maddieclayton) [inline src/ResourceManager/ContainerInstance/Commands.ContainerInstance/help/New-AzureRmContainerGroup.md]: Awesome examples! Thanks for this.

File: src/ResourceManager/ContainerInstance/Commands.ContainerInstance.Test/ScenarioTests/ContainerInstanceTests.ps1 (hunk 8 above)
Comment (yolo3301) [reply src/ResourceManager/ContainerInstance/Commands.ContainerInstance.Test/ScenarioTests/ContainerInstanceTests.ps1]: Will the session recording expose storage account key used for volume?

File: src/ResourceManager/ContainerInstance/Commands.ContainerInstance/Commands/NewAzureContainerGroupCommand.cs (hunk 10 above)
Comment (yolo3301) [reply src/ResourceManager/ContainerInstance/Commands.ContainerInstance/Commands/NewAzureContainerGroupCommand.cs]: But these are only mandatory for this parameter set. Should I still do the same?

File: src/ResourceManager/ContainerInstance/Commands.ContainerInstance/Commands/NewAzureContainerGroupCommand.cs (hunk 11 above)
Comment (yolo3301) [reply src/ResourceManager/ContainerInstance/Commands.ContainerInstance/Commands/NewAzureContainerGroupCommand.cs]: Are you suggesting to have something `PSCredential AzureFileVolumeAccountCredential` and have username and key in it?

File: src/ResourceManager/ContainerInstance/Commands.ContainerInstance/Models/PSContainerGroup.cs (hunk 13 above)
Comment (yolo3301) [reply src/ResourceManager/ContainerInstance/Commands.ContainerInstance/Models/PSContainerGroup.cs]: Yes.

File: src/ResourceManager/ContainerInstance/Commands.ContainerInstance/Models/PSContainer.cs (hunk 12 above)
Comment (yolo3301) [reply src/ResourceManager/ContainerInstance/Commands.ContainerInstance/Models/PSContainer.cs]: Previously an event belongs to a container. Now an event can also belong to the container group. Simply keep up with sdk.

File: src/ResourceManager/ContainerInstance/Commands.ContainerInstance/Commands/NewAzureContainerGroupCommand.cs (hunk 10 above)
Comment (maddieclayton) [reply src/ResourceManager/ContainerInstance/Commands.ContainerInstance/Commands/NewAzureContainerGroupCommand.cs]: Yes, they are not in the other parameter sets, so it won't affect those other parameter sets.

File: src/ResourceManager/ContainerInstance/Commands.ContainerInstance/Commands/NewAzureContainerGroupCommand.cs (hunk 11 above)
Comment (maddieclayton) [reply src/ResourceManager/ContainerInstance/Commands.ContainerInstance/Commands/NewAzureContainerGroupCommand.cs]: That's what I would consider a PSCredential to be - it requires both a username and password, so "key" seems like the wrong name.

File: src/ResourceManager/ContainerInstance/Commands.ContainerInstance.Test/ScenarioTests/ContainerInstanceTests.ps1 (hunk 8 above)
Comment (maddieclayton) [reply src/ResourceManager/ContainerInstance/Commands.ContainerInstance.Test/ScenarioTests/ContainerInstanceTests.ps1]: Yes - Can you create a test storage account and delete it at the end of the test? This should prevent any security issues caused by exposing the account key.

File: src/ResourceManager/ContainerInstance/Commands.ContainerInstance.Test/Commands.ContainerInstance.Test.csproj [hunk 15]
```diff
@@ -187,10 +187,7 @@
     <None Include="ScenarioTests\ContainerInstanceTests.ps1">
       <CopyToOutputDirectory>PreserveNewest</CopyToOutputDirectory>
     </None>
-    <None Include="SessionRecords\Microsoft.Azure.Commands.ContainerInstance.Test.ScenarioTests.ContainerInstanceTests\TestContainerGroupCommands.json">
-      <CopyToOutputDirectory>PreserveNewest</CopyToOutputDirectory>
-    </None>
-    <None Include="SessionRecords\Microsoft.Azure.Commands.ContainerInstance.Test.ScenarioTests.ContainerInstanceTests\TestContainerInstanceLog.json">
+    <None Include="SessionRecords\Microsoft.Azure.Commands.ContainerInstance.Test.ScenarioTests.ContainerInstanceTests\*">
```
Comment (maddieclayton) [inline src/ResourceManager/ContainerInstance/Commands.ContainerInstance.Test/Commands.ContainerInstance.Test.csproj]: Can you add these tests separately? We generally like to stay away from using *

File: src/ResourceManager/ContainerInstance/Commands.ContainerInstance/Commands/NewAzureContainerGroupCommand.cs [hunk 16]
```diff
@@ -60,38 +47,45 @@ public class NewAzureContainerGroupCommand : ContainerInstanceCmdletBase
...
         [Parameter(
             Mandatory = true,
-            Position = 1,
-            ParameterSetName = CreateContainerGroupWithRegistryParamSet,
-            ValueFromPipelineByPropertyName = true,
-            HelpMessage = "The container group name.")]
+            Position = 2,
+            HelpMessage = "The container image.")]
+        [ValidateNotNullOrEmpty]
+        public string Image { get; set; }
```
Comment (maddieclayton) [inline src/ResourceManager/ContainerInstance/Commands.ContainerInstance/Commands/NewAzureContainerGroupCommand.cs]: Image previously was not a named parameter - was this addition intentional? If so, that is okay, I just want to ensure that the change was intended.

**Comment by maddieclayton** [comment]
@cormacpayne Can you take a quick look to make sure this is ready to merge?

File: src/ResourceManager/ContainerInstance/Commands.ContainerInstance.Test/Commands.ContainerInstance.Test.csproj (hunk 15 above)
Comment (yolo3301) [reply src/ResourceManager/ContainerInstance/Commands.ContainerInstance.Test/Commands.ContainerInstance.Test.csproj]: Sure. Just curious why?

File: src/ResourceManager/ContainerInstance/Commands.ContainerInstance/Commands/NewAzureContainerGroupCommand.cs (hunk 16 above)
Comment (yolo3301) [reply src/ResourceManager/ContainerInstance/Commands.ContainerInstance/Commands/NewAzureContainerGroupCommand.cs]: Yes.

**Comment by maddieclayton** [comment]
@azuresdkci  test this please

**Comment by maddieclayton** [comment]
on demand run here: https://azuresdkci.westus2.cloudapp.azure.com/view/PowerShell/job/powershell-demand/
